diff --git a/mcs/tools/xbuild/ErrorUtilities.cs b/mcs/tools/xbuild/ErrorUtilities.cs
index a210486..9aed323 100644
--- a/mcs/tools/xbuild/ErrorUtilities.cs
+++ b/mcs/tools/xbuild/ErrorUtilities.cs
@@ -33,7 +33,7 @@ namespace Mono.XBuild.CommandLine {
 	public static class ErrorUtilities {
 
 		static string[] version = {
-			String.Format ("XBuild Engine Version {0}", Consts.MonoVersion),
+			String.Format ("XBuild Engine Version {0} (EventStore patched build: ThreadPool.c)", Consts.MonoVersion),
 			String.Format ("Mono, Version {0}", Consts.MonoVersion),
 			"Copyright (C) Marek Sieradzki 2005-2008, Novell 2008-2011.",
 		};
diff --git a/mono/metadata/threadpool.c b/mono/metadata/threadpool.c
index e8a2f1a..f83e473 100644
--- a/mono/metadata/threadpool.c
+++ b/mono/metadata/threadpool.c
@@ -555,8 +555,8 @@ socket_io_add (MonoAsyncResult *ares, MonoSocketAsyncResult *state)
 
 	mono_g_hash_table_replace (data->sock_to_state, state->handle, list);
 	ievt = get_events_from_list (list);
-	LeaveCriticalSection (&data->io_lock);
 	data->modify (data->event_data, fd, state->operation, ievt, is_new);
+	LeaveCriticalSection (&data->io_lock);
 }
 
 #ifndef DISABLE_SOCKETS
diff --git a/mono/mini/mini.c b/mono/mini/mini.c
index 6c9d1c0..1b4379c 100644
--- a/mono/mini/mini.c
+++ b/mono/mini/mini.c
@@ -7312,9 +7312,9 @@ char*
 mono_get_runtime_build_info (void)
 {
 	if (mono_build_date)
-		return g_strdup_printf ("%s (%s %s)", VERSION, FULL_VERSION, mono_build_date);
+		return g_strdup_printf ("%s (%s %s) (EventStore build)", VERSION, FULL_VERSION, mono_build_date);
 	else
-		return g_strdup_printf ("%s (%s)", VERSION, FULL_VERSION);
+		return g_strdup_printf ("%s (%s) (EventStore build)", VERSION, FULL_VERSION);
 }
 
 static void
